File: GUILib/src/space/kiibou/net/common/InputStreamListener.java
package space.kiibou.net.common;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.SocketException;
import java.util.Objects;
import java.util.function.Consumer;

public class InputStreamListener implements Runnable {

    private final InputStream in;
    private final Callbacks<String, Void> messageCallbacks;
    private final Callbacks<Void, Void> disconnectCallbacks;

    public InputStreamListener(final InputStream in) {
        this.in = Objects.requireNonNull(in);
        messageCallbacks = new Callbacks<>();
        disconnectCallbacks = new Callbacks<>();
    }

    @Override
    public void run() {
        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(in))) {
            while (!Thread.interrupted()) {
                final String message = reader.readLine();

                Objects.requireNonNull(message);

                messageCallbacks.callAll(message);
            }
        } catch (SocketException | NullPointerException e) {
            disconnectCallbacks.callAll(null);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public long registerMessageCallback(final Consumer<String> callback) {
        return messageCallbacks.addCallback(message -> {
            callback.accept(message);
            return null;
        });
    }

    public long registerDisconnectCallback(final Runnable callback) {
        return disconnectCallbacks.addCallback(ignore -> {
            callback.run();
            return null;
        });
    }

    public void stop() throws IOException {
        Thread.currentThread().interrupt();
        in.close();
    }
}

File: src/serverMain/kotlin/space/kiibou/net/common/InputStreamListener.kt
package space.kiibou.net.common

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.SocketException

class InputStreamListener(private val input: InputStream) : Runnable {
    private val messageCallbacks: Callbacks<String, Unit> = Callbacks()
    private val disconnectCallbacks: Callbacks<Unit, Unit> = Callbacks()

    override fun run() {
        try {
            BufferedReader(InputStreamReader(input)).use { reader ->
                while (!Thread.interrupted()) {
                    val message = reader.readLine()
                    messageCallbacks.callAll(message)
                }
            }
        } catch (e: SocketException) {
            disconnectCallbacks.callAll(Unit)
        } catch (e: NullPointerException) {
            disconnectCallbacks.callAll(Unit)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun registerMessageCallback(callback: (String) -> Unit): Long {
        return messageCallbacks.addCallback(callback)
    }

    fun registerDisconnectCallback(callback: (Unit) -> Unit): Long {
        return disconnectCallbacks.addCallback(callback)
    }

    fun stop() {
        Thread.currentThread().interrupt()
        input.close()
    }
}
